
# Basic Overview
The objective is to explore h2o package in python and see how we can make randomForest models using the same.

Comments/criticisms/appreciations are greatly accepted and appreciated. Do not be shy and send me an email at babinu@gmail.com !

Source of data : https://www.kaggle.com/c/titanic/data

### A note about randoForest and h2o

This is more of binary classification problem and hence random forest appears to be a natural first choice.

However, we cannot yet use categorical predictors correctly in scikit learn (https://github.com/scikit-learn/scikit-learn/pull/4899) and hence we use another package named h2o to do the same(http://docs.h2o.ai/h2o/latest-stable/h2o-py/docs/intro.html)

A more detailed description can be seen here : https://roamanalytics.com/2016/10/28/are-categorical-variables-getting-lost-in-your-random-forests/

In [1]:
import pandas as pd
import numpy as np

In [2]:
train_data = pd.read_csv("../train.csv")
test_data = pd.read_csv("../test.csv")

In [3]:
len(train_data)

891

In [4]:
train_data['Predictions'] = -1

In [5]:
# Using h2o
import h2o
h2o.init(nthreads = -1, max_mem_size = 8)

Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  ----------------------------------------
H2O cluster uptime:         3 days 23 hours 18 mins
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.18.0.2
H2O cluster version age:    3 months and 20 days !!!
H2O cluster name:           H2O_from_python_babs4JESUS_c6btew
H2O cluster total nodes:    1
H2O cluster free memory:    6.770 Gb
H2O cluster total cores:    8
H2O cluster allowed cores:  8
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.4 final
--------------------------  ----------------------------------------

In [6]:
def get_h2o_frame_with_rel_factors_test(clean_training_data_v3):
    clean_h2o_data = h2o.H2OFrame(clean_training_data_v3)    
    clean_h2o_data['Sex'] = clean_h2o_data['Sex'].asfactor()
    clean_h2o_data['Pclass'] = clean_h2o_data['Pclass'].asfactor()
    return clean_h2o_data

In [7]:
def get_h2o_frame_with_rel_factors(clean_training_data_v3):
    
    clean_h2o_data = get_h2o_frame_with_rel_factors_test(clean_training_data_v3)
    clean_h2o_data['Survived'] = clean_h2o_data['Survived'].asfactor()
    return clean_h2o_data

In [8]:
train_data_h2o = get_h2o_frame_with_rel_factors(train_data)
test_data_h2o = get_h2o_frame_with_rel_factors_test(test_data)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [9]:
# Split to 70% training, 15% validation and 15% for test. Seed set to fixed value to ensure reproducibility.
splits = train_data_h2o.split_frame(ratios=[0.75], seed=1)  

train_validation = splits[0]
test = splits[1]

In [10]:
y_column = 'Survived'
x_columns = 'Sex'

In [11]:
# Import H2O RF:
from h2o.estimators.random_forest import H2ORandomForestEstimator


In [12]:
# Do a 10 fold cross validation as that is done typically.
rf_fit_cross_val = H2ORandomForestEstimator(model_id='rf_fit_cross_val', seed=1, nfolds=10)
rf_fit_cross_val.train(x=x_columns, y=y_column, 
                       training_frame=train_validation)


drf Model Build progress: |███████████████████████████████████████████████| 100%


### Getting useful metrics.
Now, we will see as to how to extract useful metrics from this cross validated model object.

In [13]:
# Cross validated auc(). Details can be found here : 
# http://docs.h2o.ai/h2o/latest-stable/h2o-docs/cross-validation.html
print("Cross validated auc () is ", format(rf_fit_cross_val.auc(xval=True), '0.4f'))

# AUC() directly obtained from training data
print("AUC ()  obtained directly from training data is ", format(rf_fit_cross_val.auc(), '0.4f'))


Cross validated auc () is  0.7286
AUC ()  obtained directly from training data is  0.7646


### Dissecting AUC

What exactly is this AUC() and how do we interpret this result ?

A detailed explanation can be found in https://stats.stackexchange.com/questions/145566/how-to-calculate-area-under-the-curve-auc-or-the-c-statistic-by-hand.

For now, I will explaining as to how we got the computed value by doing a paralled computation by hand.

In [14]:
rf_fit_cross_val

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  rf_fit_cross_val


ModelMetricsBinomial: drf
** Reported on train data. **

MSE: 0.1631173476185785
RMSE: 0.4038778870136102
LogLoss: 0.5065569105131861
Mean Per-Class Error: 0.2316714975845411
AUC: 0.7645845410628019
Gini: 0.5291690821256039
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.7393161333524264: 


0    1    Error    Rate
-----  ---  ---  -------  -------------
0      353  61   0.1473   (61.0/414.0)
1      79   171  0.316    (79.0/250.0)
Total  432  232  0.2108   (140.0/664.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.739316     0.714876  11
max f2                       0.17907      0.751202  27
max f0point5                 0.739316     0.729342  11
max accuracy                 0.739316     0.792169  11
max precision                0.739316     0.758242  1
max recall                   0.17907      1         27
max specificity              0.739316     0.958937  0
max absolute_mcc             0.739316     0.552397  11
max min_per_class_accuracy   0.17907      0.72      14
max mean_per_class_accuracy  0.739316     0.772329  11

Gains/Lift Table: Avg response rate: 37.65 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0978916                   0.739316           1.96135   1.96135            0.738462         0.738462                    0.192           0.192                      96.1354   96.1354
    2        0.137048                    0.739316           2.14523   2.01389            0.807692         0.758242                    0.084           0.276                      114.523   101.389
    3        0.168675                    0.739316           1.77067   1.96829            0.666667         0.741071                    0.056           0.332                      77.0667   96.8286
    4        0.215361                    0.739316           1.97058   1.96878            0.741935         0.741259                    0.092           0.424                      97.0581   96.8783
    5        0.307229                    0.739316           1.9158    1.95294            0.721311         0.735294                    0.176           0.6                        91.5803   95.2941
    6        0.427711                    0.17907            0.996     1.68338            0.375            0.633803                    0.12            0.72                       -0.4      68.338
    7        0.537651                    0.17907            0.363836  1.41356            0.136986         0.532213                    0.04            0.76                       -63.6164  41.3557
    8        0.599398                    0.17907            0.647805  1.33467            0.243902         0.502513                    0.04            0.8                        -35.2195  33.4673
    9        0.701807                    0.17907            0.507765  1.21401            0.191176         0.457082                    0.052           0.852                      -49.2235  21.4009
    10       0.843373                    0.17907            0.48034   1.09086            0.180851         0.410714                    0.068           0.92                       -51.966   9.08571
    11       0.917169                    0.17907            0.487837  1.04234            0.183673         0.392447                    0.036           0.956                      -51.2163  4.23383
    12       1                           0.17907            0.5312    1                  0.2              0.376506                    0.044           1                          -46.88    0



ModelMetricsBinomial: drf
** Reported on cross-validation data. **

MSE: 0.1643192354001771
RMSE: 0.4053630908212748
LogLoss: 0.510189193374565
Mean Per-Class Error: 0.22767149758454108
AUC: 0.7285507246376812
Gini: 0.45710144927536245
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.7156862616539001: 


0    1    Error    Rate
-----  ---  ---  -------  -------------
0      353  61   0.1473   (61.0/414.0)
1      77   173  0.308    (77.0/250.0)
Total  430  234  0.2078   (138.0/664.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.715686     0.714876  8
max f2                       0.166667     0.751202  18
max f0point5                 0.715686     0.729342  8
max accuracy                 0.715686     0.792169  8
max precision                0.715686     0.739316  8
max recall                   0.166667     1         18
max specificity              0.75         0.983092  0
max absolute_mcc             0.715686     0.552397  8
max min_per_class_accuracy   0.188329     0.716     9
max mean_per_class_accuracy  0.715686     0.772329  8

Gains/Lift Table: Avg response rate: 37.65 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0271084                   0.75               1.62311   1.62311            0.611111         0.611111                    0.044           0.044                      62.3111   62.3111
    2        0.0557229                   0.748837           1.67747   1.65103            0.631579         0.621622                    0.048           0.092                      67.7474   65.1027
    3        0.126506                    0.741463           1.86485   1.77067            0.702128         0.666667                    0.132           0.224                      86.4851   77.0667
    4        0.204819                    0.740385           1.94092   1.83576            0.730769         0.691176                    0.152           0.376                      94.0923   83.5765
    5        0.307229                    0.734884           2.03106   1.90086            0.764706         0.715686                    0.208           0.584                      103.106   90.0863
    6        0.432229                    0.188329           1.056     1.65653            0.39759          0.623693                    0.132           0.716                      5.6       65.653
    7        0.560241                    0.185567           0.312471  1.34942            0.117647         0.508065                    0.04            0.756                      -68.7529  34.9419
    8        0.620482                    0.184615           0.332     1.25064            0.125            0.470874                    0.02            0.776                      -66.8     25.0641
    9        0.701807                    0.183511           0.393481  1.15131            0.148148         0.433476                    0.032           0.808                      -60.6519  15.1313
    10       0.820783                    0.173469           0.537924  1.0624             0.202532         0.4                         0.064           0.872                      -46.2076  6.24
    11       0.948795                    0.170918           0.656188  1.00759            0.247059         0.379365                    0.084           0.956                      -34.3812  0.759365
    12       1                           0.166667           0.859294  1                  0.323529         0.376506                    0.044           1                          -14.0706  0


Cross-Validation Metrics Summary: 


mean      sd         cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid    cv_8_valid    cv_9_valid    cv_10_valid
-----------------------  --------  ---------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  -------------
accuracy                 0.789515  0.0258059  0.835443      0.807229      0.803279      0.779661      0.823529      0.711864      0.754386      0.826087      0.760563      0.793103
auc                      0.77136   0.0266816  0.815185      0.787995      0.773396      0.740789      0.828125      0.713218      0.7375        0.818056      0.738916      0.760417
err                      0.210485  0.0258059  0.164557      0.192771      0.196721      0.220339      0.176471      0.288136      0.245614      0.173913      0.239437      0.206897
err_count                13.8      1.40712    13            16            12            13            12            17            14            12            17            12
f0point5                 0.719099  0.0451879  0.736434      0.724138      0.645161      0.659341      0.865385      0.730769      0.742574      0.742188      0.72          0.625
f1                       0.706176  0.036876   0.745098      0.724138      0.666667      0.648649      0.818182      0.690909      0.681818      0.76          0.679245      0.647059
f2                       0.695962  0.038852   0.753968      0.724138      0.689655      0.638298      0.775862      0.655172      0.630252      0.778689      0.642857      0.670732
lift_top_group           1.98653   0.17939    2.30923       2.07253       2.26625       2.07018       1.7           1.49467       1.8           2.10096       1.83621       2.21528
logloss                  0.514018  0.0407193  0.442727      0.481849      0.478013      0.516694      0.501408      0.639237      0.576227      0.454814      0.558534      0.490678
max_per_class_error      0.309521  0.0446014  0.24          0.275862      0.294118      0.368421      0.25          0.366667      0.4           0.208333      0.37931       0.3125
mcc                      0.543404  0.0495933  0.623933      0.57599       0.529384      0.488704      0.659695      0.431425      0.5           0.625207      0.496529      0.503173
mean_per_class_accuracy  0.77136   0.0266816  0.815185      0.787995      0.773396      0.740789      0.828125      0.713218      0.7375        0.818056      0.738916      0.760417
mean_per_class_error     0.22864   0.0266816  0.184815      0.212005      0.226604      0.259211      0.171875      0.286782      0.2625        0.181944      0.261084      0.239583
mse                      0.165947  0.0168824  0.135919      0.152821      0.152015      0.168133      0.157944      0.217817      0.191338      0.141242      0.184619      0.157623
precision                0.729451  0.0554987  0.730769      0.724138      0.631579      0.666667      0.9           0.76          0.789474      0.730769      0.75          0.611111
r2                       0.271472  0.0563817  0.37165       0.327724      0.243788      0.229906      0.366031      0.128483      0.22293       0.377358      0.235907      0.210945
recall                   0.690479  0.0446014  0.76          0.724138      0.705882      0.631579      0.75          0.633333      0.6           0.791667      0.62069       0.6875
rmse                     0.406359  0.0202404  0.368673      0.390923      0.389891      0.41004       0.397421      0.466708      0.437422      0.375822      0.429673      0.397018
specificity              0.852241  0.0197949  0.87037       0.851852      0.840909      0.85          0.90625       0.793103      0.875         0.844444      0.857143      0.833333

Scoring History: 


timestamp            duration    number_of_trees    training_rmse        training_logloss    training_auc        training_lift       training_classification_error
---  -------------------  ----------  -----------------  -------------------  ------------------  ------------------  ------------------  -------------------------------
     2018-06-26 13:37:00  0.340 sec   0.0                nan                  nan                 nan                 nan                 nan
     2018-06-26 13:37:00  0.341 sec   1.0                0.4182648963938531   0.5355520466612966  0.7601503759398496  2.0405853658536586  0.2236842105263158
     2018-06-26 13:37:00  0.342 sec   2.0                0.40744197502405993  0.5145611728357857  0.7747844827586207  2.057464788732394   0.20918367346938777
     2018-06-26 13:37:00  0.343 sec   3.0                0.4070305345199281   0.5128466868991368  0.7711684096646232  1.9843678160919538  0.2111801242236025
     2018-06-26 13:37:00  0.343 sec   4.0                0.40791167779024345  0.5143936744665389  0.7694027667488611  1.9718787878787878  0.21272727272727274
---  ---                  ---         ---                ---                  ---                 ---                 ---                 ---
     2018-06-26 13:37:00  0.406 sec   46.0               0.40387788673893427  0.5065569096196756  0.7751980676328503  1.9584646464646462  0.20783132530120482
     2018-06-26 13:37:00  0.408 sec   47.0               0.40387788662830976  0.5065569094175242  0.7754879227053141  1.999460674157303   0.20783132530120482
     2018-06-26 13:37:00  0.410 sec   48.0               0.4038778869518281   0.5065569102867875  0.7685265700483093  2.009025641025641   0.20783132530120482
     2018-06-26 13:37:00  0.411 sec   49.0               0.4038778871475183   0.5065569108384887  0.7620144927536232  1.9826478873239435  0.20783132530120482
     2018-06-26 13:37:00  0.413 sec   50.0               0.4038778870136102   0.5065569105131861  0.7645845410628019  1.9613538461538462  0.20783132530120482


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
Sex         2378.18                1                    1

NOTE : An issue with the confusion matrix has been logged in https://github.com/h2oai/h2o-tutorials/issues/71

We compute the AUC by hand, by computing the area in the ROC curve.

The ROC curve is computed by finding the relevant data points.

Data point #1 -> Corresponding to sensitivity = 0 and specificity = 1 (or (1 - specificity) = 0)
Data point #2 -> Corresopnding to sensitivity = 1 and specificity = 0 (or (1 - specificity) = 1)
Data point #3 -> Corresponding to the case we have our predictions as 1 and see how many survived.

In this scenario => When we predict as survived, we see that we obtained 173 out of 250 cases => Sensitivity = 173/250
Also, we see that in we actually predict 353 (out of the total 414 cases) of not survived correctly => Specificity = 353/432

=> Our relevant data point = (1- 353/432, 173/250)
=> AUC = Area under ROC curve => Area of the triangle and relevant trapeziod , which when computed turns out to be = 0.5 * ( 353/414 + 173/250 ) = 0.77232850241

NOTE : The expression for area, comes out nicely as 0.5*(sensitivity + specificity).

Let us compare with the value of AUC computed from the model directly .

In [15]:
perf = rf_fit_cross_val.model_performance()  #train=True is the default, so it's not needed
perf.plot() 
perf.show()


ModelMetricsBinomial: drf
** Reported on train data. **

MSE: 0.1631173476185785
RMSE: 0.4038778870136102
LogLoss: 0.5065569105131861
Mean Per-Class Error: 0.2316714975845411
AUC: 0.7645845410628019
Gini: 0.5291690821256039
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.7393161333524264: 


0    1    Error    Rate
-----  ---  ---  -------  -------------
0      353  61   0.1473   (61.0/414.0)
1      79   171  0.316    (79.0/250.0)
Total  432  232  0.2108   (140.0/664.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.739316     0.714876  11
max f2                       0.17907      0.751202  27
max f0point5                 0.739316     0.729342  11
max accuracy                 0.739316     0.792169  11
max precision                0.739316     0.758242  1
max recall                   0.17907      1         27
max specificity              0.739316     0.958937  0
max absolute_mcc             0.739316     0.552397  11
max min_per_class_accuracy   0.17907      0.72      14
max mean_per_class_accuracy  0.739316     0.772329  11

Gains/Lift Table: Avg response rate: 37.65 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0978916                   0.739316           1.96135   1.96135            0.738462         0.738462                    0.192           0.192                      96.1354   96.1354
    2        0.137048                    0.739316           2.14523   2.01389            0.807692         0.758242                    0.084           0.276                      114.523   101.389
    3        0.168675                    0.739316           1.77067   1.96829            0.666667         0.741071                    0.056           0.332                      77.0667   96.8286
    4        0.215361                    0.739316           1.97058   1.96878            0.741935         0.741259                    0.092           0.424                      97.0581   96.8783
    5        0.307229                    0.739316           1.9158    1.95294            0.721311         0.735294                    0.176           0.6                        91.5803   95.2941
    6        0.427711                    0.17907            0.996     1.68338            0.375            0.633803                    0.12            0.72                       -0.4      68.338
    7        0.537651                    0.17907            0.363836  1.41356            0.136986         0.532213                    0.04            0.76                       -63.6164  41.3557
    8        0.599398                    0.17907            0.647805  1.33467            0.243902         0.502513                    0.04            0.8                        -35.2195  33.4673
    9        0.701807                    0.17907            0.507765  1.21401            0.191176         0.457082                    0.052           0.852                      -49.2235  21.4009
    10       0.843373                    0.17907            0.48034   1.09086            0.180851         0.410714                    0.068           0.92                       -51.966   9.08571
    11       0.917169                    0.17907            0.487837  1.04234            0.183673         0.392447                    0.036           0.956                      -51.2163  4.23383
    12       1                           0.17907            0.5312    1                  0.2              0.376506                    0.044           1                          -46.88    0

How do we explain the mismatch between the AUC numbers ?

I am not very sure, from the the following link(http://gim.unmc.edu/dxtests/roc3.htm) and from the plot obtained from the summary, it looks like h2o computes AUC using a parametric method, whereas we do an approximation using trapezoids.

Moreover, when we compute the metric using model performance function and a relevant function in scikit learn, the values match exactly. 

In [16]:
# Using model performance function.
print("AUC ()  obtained directly from training data is ", 
      format(rf_fit_cross_val.model_performance(train_validation).auc(), '0.10f')) 

AUC ()  obtained directly from training data is  0.7723285024


In [17]:
# Using relevant routine from scikit learn.
survived_set = rf_fit_cross_val.predict(train_validation)['predict']
survived_set['Survived'] = train_validation['Survived']
survived_set_df = survived_set.as_data_frame()
from sklearn.metrics import roc_curve, auc, roc_auc_score

fp_rate, tp_rate, thresholds = roc_curve(survived_set_df['Survived'], survived_set_df['predict'])
print(auc(fp_rate, tp_rate))
print(roc_auc_score(survived_set_df['Survived'], survived_set_df['predict']))

drf prediction progress: |████████████████████████████████████████████████| 100%
0.7723285024154589
0.7723285024154589


### What are the takeaways from this ?

1. Though we understand the basic computation of AUC, we are not able to replicate exactly as to how it is computed by the h2o model internally. This will make it tricky when we build complex models later and will be forced to compute auc by hand and compare with earlier values.

2. We are able compute auc by hand using the trapezoid approximation and are able to match up with values generated by scikit learn. We can opt to use this method of computing auc, rather than the one implicitly used by the model.

3. We can ignore auc and use a fit score, measuring the number of observations we are able to predict correctly and use that to refine our model. 

Practically, I believe we can go for both #2 and #3. Let us see how it goes !

In [18]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Predictions  
0      0         A/5 21171   7.2500   NaN        S           -1  
1      0          PC 17599  71.2833   C85        C           -1  
2      0  STON/O2. 3101282   7.9250   NaN        S           -1  
3      0            113803  53.1000  C123        S           -1  
4      0            373450   8.0500   NaN        S           -1

In [19]:
# Function for splitting input dataframe into training/validation and test sets.
def get_train_validation_test_data(input_df, ratio=0.75):
    train_len = int(ratio * len(input_df))
    train_validation = input_df[:train_len]
    test = input_df[train_len:]
    return (train_validation, test)


In [20]:
(train_validation, test) = get_train_validation_test_data(train_data)
test_h2o = get_h2o_frame_with_rel_factors_test(test)
train_validation_h2o = get_h2o_frame_with_rel_factors(train_validation)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [21]:
def get_predictions_given_model_and_data(rf_fit, train_validation, train_validation_h2o):
    # Using relevant routine from scikit learn.
    predict_out = rf_fit.predict(train_validation_h2o) 
    if 'Survived' in train_validation.columns:
        survived_set = pd.DataFrame(columns=['Survived', 'Predictions'])
        survived_set['Survived'] = train_validation['Survived']
    else:
        survived_set = pd.DataFrame(columns=['Predictions'])
    survived_set['Predictions'] = predict_out.as_data_frame()['predict'].values.tolist()
    
    return survived_set

In [22]:
# The bread and butter routine for cross validation.
def get_cross_val_output(input_df, x_columns=['Sex'], y_column = 'Survived', nfolds=10):
    partition_indices = np.array_split(np.arange(len(input_df)), nfolds)
    
    cross_validated_data = pd.DataFrame(columns=input_df.columns)
    for i in range(nfolds):
        cross_validated_set = input_df[partition_indices[i][0]:partition_indices[i][-1] + 1].copy()
        rel_training_data = pd.DataFrame(columns=input_df.columns)
        for j in range(nfolds):
            if j != i:
                training_set = input_df[partition_indices[j][0]:partition_indices[j][-1] + 1]
                rel_training_data = pd.concat([rel_training_data, training_set])
        h2o_train_data = get_h2o_frame_with_rel_factors(rel_training_data)

        rf_fit = H2ORandomForestEstimator(model_id='rf_fit', seed=1)
        rf_fit.train(x=x_columns, y=y_column, training_frame=h2o_train_data)

        predict_out = rf_fit.predict(
            get_h2o_frame_with_rel_factors_test(cross_validated_set))                

        cross_validated_set['Predictions'] = predict_out.as_data_frame()['predict'].values.tolist()
        cross_validated_data = pd.concat([cross_validated_data, cross_validated_set])

    return cross_validated_data
    

In [23]:
cross_validated_output = get_cross_val_output(train_validation, x_columns=['Sex'], y_column = 'Survived', nfolds=10)

Parse progress: |█████████████████████████████████████████████████████████| 100%
drf Model Build progress: |███████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
drf prediction progress: |████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
drf Model Build progress: |███████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
drf prediction progress: |████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
drf Model Build progress: |███████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
drf prediction progress: |████████████████████████████████████████████████| 100%
Parse progress: |███████████

In [24]:
assert(cross_validated_output[cross_validated_output['Predictions'] == -1].empty)

In [25]:
print("Fit score obtained through cross validation is ", 
      np.sum((cross_validated_output['Survived'] == cross_validated_output['Predictions'])/len(
          cross_validated_output['Survived'])))

Fit score obtained through cross validation is  0.7859281437125747


In [26]:
print(roc_auc_score(cross_validated_output['Survived'].apply(lambda x : int(x)), 
                    cross_validated_output['Predictions'].apply(lambda x : int(x))))

0.768947963800905


### Out of sample testing.
Let us see how out of sample testing looks with this model. Intuitively, one would not expect the results to be much different from that of cross validation.

In [27]:

rf_fit_cross_val.model_performance(test_h2o).auc()

0.7593409444732745

In [28]:
rf_fit_cross_val.model_performance(train_validation_h2o).auc()
rf_fit_cross_val.auc(xval=True)

0.7285507246376812

Let us see how things look, using the performance metrics designed by us 

In [29]:
predictions_set = get_predictions_given_model_and_data(rf_fit_cross_val, test, test_h2o)

drf prediction progress: |████████████████████████████████████████████████| 100%


In [30]:
print("Computed AUC is ", 
      roc_auc_score(predictions_set['Survived'].apply(lambda x : int(x)), 
                    predictions_set['Predictions'].apply(lambda x : int(x))))

Computed AUC is  0.7593409444732746


In [31]:
print("Computed fit score is ",
      np.sum((predictions_set['Survived'] == predictions_set['Predictions'])/len(
          predictions_set['Survived'])))

Computed fit score is  0.7892376681614349


### Train the model on entire training data

In [32]:
model1_full_train_data = H2ORandomForestEstimator(model_id='model1_full_train_data', seed=1)
model1_full_train_data.train(x=x_columns, y=y_column, 
                             training_frame=train_data_h2o)


drf Model Build progress: |███████████████████████████████████████████████| 100%


###  Making predictions
Let us make predictions on the test data here !

In [33]:
predict_out = get_predictions_given_model_and_data(model1_full_train_data, test_data, test_data_h2o)

drf prediction progress: |████████████████████████████████████████████████| 100%


In [34]:
test_data['Survived'] = predict_out['Predictions']

Some sanity on test data !

In [35]:
test_data[(test_data['Survived'] ==1) &(test_data['Sex'] == 'male')]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked, Survived]
Index: []

In [36]:
test_data[['PassengerId', 'Survived']].sort_values(by=['PassengerId']).to_csv("output_only_gender.csv", index=False)
# Kaggle score : 0.76555 ???